flax/testing/benchmark.py

```diff
@@ -0,0 +1,239 @@
+    """Report a single extra for the benchmark."""
+    self.report_extras({name: value})
+
+  def _get_test_name(self, prefix='test_'):
+    """Returns full name of test class and method calling report_benchmark.
+
+    The name is based on the *outermost* Benchmark class in the class stack.
+    Based on tensorflow/python/platform/benchmark.py
+
+    Args:
+      prefix: str. Prefix that the caller method must have.
+
+    Returns:
+      Resolved test name as `ClassName.test_name`.
+    """
+    # Find the caller method (outermost Benchmark class).
+    stack = inspect.stack()
+    calling_class, name = None, None
+    for frame_info in stack[::-1]:
+      f_locals = frame_info.frame.f_locals
+      f_self = f_locals.get('self', None)
+      if isinstance(f_self, Benchmark):
+        name = frame_info.function
+        if name.startswith(prefix):
+          calling_class = f_self
+          break
+    if calling_class is None:
+      raise ValueError('Unable to determine the calling Benchmark class.')
+
+    # Prefix the name with the class name.
+    class_name = type(calling_class).__name__
+    name = "%s.%s" % (class_name, name)
+    return name
+
+  def _update_reported_name(self):
+    """Record / update test name for the benchmark."""
+    self._reported_name = self._reported_name or self._get_test_name()
+
+  def _report_benchmark_results(self):
+    """Produce benchmark results report.
+
+    Results are reported as a JSON string with the following schema:
+    ```
+    {
+      "name": <class.testMethod>
+      "succeeded": true / false
+      "wall_time": float (containing wall-time for the benchmark)
+      "metrics": {
+        "string" -> float map of other performance metrics
+      }
+      "extras": {
+        "string" -> "string" map containing anything else of interest
+      }
+    }
+    ```
+    """
+    name = self._reported_name
+    if not name:
+      raise ValueError('Unable to determine test name for reporting '
+                       'benchmark results. Make sure you are using '
+                       '`self.report_*` methods.')
+
+    succeeded = not self.has_outstanding_fails()
+    results = {'name': name,
+               'succeeded': succeeded,
+               'metrics': self._reported_metrics,
+               'extras': self._reported_extras}
+    if self._reported_wall_time is not None:
+      results['wall_time'] = self._reported_wall_time
+    if not succeeded:
+      msg = '\n'.join([str(fail) for fail in self._outstanding_fails])
+      results['extras']['failed_assertions'] = msg
+
+    results_str = json.dumps(results)
+    logging.info(results_str)
+
+    # Maybe save results as a file for pickup by CI / monitornig frameworks.
+    benchmark_output_dir = FLAGS.benchmark_output_dir
+    if benchmark_output_dir:
+      os.makedirs(benchmark_output_dir, exist_ok=True)
+      filename = os.path.join(benchmark_output_dir, name + '.json')
+      with open(filename, 'w') as fout:
```

Comment:
I think we'll need to use`tf.gfile` (or if there's something equivalent that doesn't depend on TF) to allow writing files to network file systems as well such as Google Cloud Storage (`gs://`) or others.

But we can cross that bridge when we get there.

Reply:
Yes. This crossed my mind. I tried quickly searching for a gfile implementation that is not bound to tf, and I couldn't find one. I wanted to avoid creating a dependency on TensorFlow just to be able to use gfile. But not sure what other alternatives we have.

Reply:
@mohitreddy1996, I'm curious -- do you know of something open source like tf.io.gfile (that supports network file systems natively) that's not bound to TF?

Reply:
Hi Avital and Alexey, 

After filing [Issue 210](https://github.com/google/flax/issues/210), I started investigating if tf.gfile could be replaced. I found the following alternatives, may be we could discuss more on how they could be useful or used as alternatives:

- [Dask](https://docs.dask.org/en/latest/remote-data-services.html#remote-data) has an IO library which allows reading/writing to local, network FS and cloud object stores (gcs, s3). Using this would mean creating a dependency on Dask (which I am not sure is a good idea as we are trying to get rid of an external dependency already - tf). 
- [fsspec](https://filesystem-spec.readthedocs.io/en/latest/intro.html) supports local, network FS, cloud object stores and even reading from Github. Although the [built-in implementations](https://filesystem-spec.readthedocs.io/en/latest/intro.html) doesn't list GCS, but looking at the [implementation](https://github.com/intake/filesystem_spec/blob/de54f06c08989e902e1703621ed99b15feac4085/fsspec/registry.py#L31), I believe it should be possible.
- [GCSFS](https://gcsfs.readthedocs.io/en/latest/): Part of Dask, but can be used as a standalone module. I haven't seen any documentation which says 'local' FS is supported (which means we will need to implement additional utility module to handle cases), but I haven't tried it out yet. Also the usage is not really extendable i.e doesn't support other cloud storages etc.  
- [Pyfilesystems](https://docs.pyfilesystem.org/en/latest/index.html) - Has a support for GCS from a [third party implementation](https://www.pyfilesystem.org/page/index-of-filesystems/). Also fsspec [claims to be better](https://filesystem-spec.readthedocs.io/en/latest/intro.html#not-pyfilesystems) than this.

Other FS interfaces rejected:
- [Apache Arrow's FS Interface](https://arrow.apache.org/docs/python/filesystems.html) - dependency on Apache Arrow also doesn't support GCS.

Reply:
Hi Mohit,

Thank you for looking into this - this is super useful; and these look like viable leads. However, just keep things easy and pragmatic, perhaps we should keep using `os` and `open` for the moment. And once we find that this is insufficient, we should probably just switch to `tf.gfile` (thus  introducing the dependency), and later figure out how to avoid the dependency.

I am also now thinking that perhaps depending on TensorFlow isn't too bad. We already have to depend on it through `tensorflow_datasets` (which is typically used in the examples), so in most cases depending on `tf.gfile` will not introduce a new dependency.

Reply:
Yup that sounds good as well :)